lib/PuppeteerSharp/Helpers/TaskHelper.cs

```diff
@@ -43,9 +47,78 @@ public static async Task WithTimeout(this Task task, int milliseconds = 1_000)
         /// </summary>
         /// <returns>The task result.</returns>
         /// <param name="task">Task to wait for.</param>
+        /// <param name="timeoutAction">Action to be executed on Timeout.</param>
         /// <param name="milliseconds">Milliseconds timeout.</param>
+        public static async Task WithTimeout(
+            this Task task,
+            Func<Task> timeoutAction,
+            int milliseconds = 1_000)
+        {
+            var tcs = new TaskCompletionSource<bool>();
```

Comment:
maybe create a private method for creating the `TaskCompletionSource` & `CancellationTokenSource` and canceling after the timeout? this code is duplicated multiple times
something like:
```cs
private static (TaskCompletionSource<bool> tcs, CancellationTokenSource cancellationSource) CreateSources(int milliseconds = 1_000)
{
    var tcs = new TaskCompletionSource<bool>();
    var cancellationSource = new CancellationTokenSource();
    if (milliseconds > 0)
    {
        cancellationToken.CancelAfter(milliseconds);
    }
    return (tcs, cancellationSource);
}
```